### ISSUE TITLE
VM::init_loadpath won't work with static builds

### ISSUE THREAD

**Issue by danielpclark** [danielpclark/rutie#45]
When choosing to compile with `RUBY_STATIC` flag the test for `class::vm::VM::init_loadpath` fails.

**Comment by danielpclark** [comment]
Fixing this may fix #44

**Comment by felix-d** [comment]
This is the failures I get when running tests on my machine with a static build.

```
failures:

---- src/class/string.rs - class::string::RString::is_valid_encoding (line 530) stdout ----
thread 'src/class/string.rs - class::string::RString::is_valid_encoding (line 530)' panicked at 'test executable failed:

dyld: lazy symbol binding failed: Symbol not found: _rb_encdb_declare
  Referenced from: /opt/rubies/ruby-2.5.3/lib/ruby/2.5.0/x86_64-darwin18/enc/encdb.bundle
  Expected in: flat namespace

dyld: Symbol not found: _rb_encdb_declare
  Referenced from: /opt/rubies/ruby-2.5.3/lib/ruby/2.5.0/x86_64-darwin18/enc/encdb.bundle
  Expected in: flat namespace


', librustdoc/test.rs:367:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- src/class/string.rs - class::string::RString::from_bytes (line 110) stdout ----
thread 'src/class/string.rs - class::string::RString::from_bytes (line 110)' panicked at 'test executable failed:

dyld: lazy symbol binding failed: Symbol not found: _rb_encdb_declare
  Referenced from: /opt/rubies/ruby-2.5.3/lib/ruby/2.5.0/x86_64-darwin18/enc/encdb.bundle
  Expected in: flat namespace

dyld: Symbol not found: _rb_encdb_declare
  Referenced from: /opt/rubies/ruby-2.5.3/lib/ruby/2.5.0/x86_64-darwin18/enc/encdb.bundle
  Expected in: flat namespace


', librustdoc/test.rs:367:17

---- src/class/vm.rs - class::vm::VM::init_loadpath (line 53) stdout ----
thread 'src/class/vm.rs - class::vm::VM::init_loadpath (line 53)' panicked at 'test executable failed:

dyld: lazy symbol binding failed: Symbol not found: _rb_encdb_declare
  Referenced from: /opt/rubies/ruby-2.5.3/lib/ruby/2.5.0/x86_64-darwin18/enc/encdb.bundle
  Expected in: flat namespace

dyld: Symbol not found: _rb_encdb_declare
  Referenced from: /opt/rubies/ruby-2.5.3/lib/ruby/2.5.0/x86_64-darwin18/enc/encdb.bundle
  Expected in: flat namespace


', librustdoc/test.rs:367:17


failures:
    src/class/string.rs - class::string::RString::from_bytes (line 110)
    src/class/string.rs - class::string::RString::is_valid_encoding (line 530)
    src/class/vm.rs - class::vm::VM::init_loadpath (line 53)

test result: FAILED. 181 passed; 3 failed; 3 ignored; 0 measured; 0 filtered out
```

**Comment by felix-d** [comment]
The problem is not related to the particular missing symbol mentioned in the stacktrace above.
Actually, any `VM::require` call to require a `.bundle` (osx) or `.so` (linux) will raise an exception for missing symbols, which live in libruby.

**Comment by danielpclark** [comment]
The Ruby programming language's source code decides which symbol names to export publicly and which ones remain private.  Sometimes these are even OS specific as can be seen from #49 .

`rb_encdb_declare` is defined here: [Ruby - ruby/encoding.c#L353](https://github.com/ruby/ruby/blob/4e62527c3306268f9f9afbcb9639048edc9069f0/encoding.c#L353) and used in the internal header here [Ruby - ruby/internal.h#L1410](https://github.com/ruby/ruby/blob/c6ef5bc31e547ee905e5a218d293fef61647f65c/internal.h#L1410) which means that we won't have access to it.  The text encoding database is either available to be included, or included, by calling `ruby_init_loadpath` which is available publicly and that method calls out to `ruby_init_loadpath_safe` here [Ruby - ruby/ruby.c#L562](https://github.com/ruby/ruby/blob/b756b25c60a887d5d073db7ccbb5e214b5cfd939/ruby.c#L562) in which you can see has many C preprocessor conditionals which I suspect is causing this issue for static linking for us.

Without the `ruby_init_loadpath` giving us access to the source code available from [Ruby - ruby/encoding.c](https://github.com/ruby/ruby/blob/4e62527c3306268f9f9afbcb9639048edc9069f0/encoding.c#L353) we have only very limited use of text encodings which is also somewhat OS dependent and I believe typically includes the following from `miniinit.c`.

https://github.com/ruby/ruby/blob/4444025d16ae1a586eee6a0ac9bdd09e33833f3c/miniinit.c#L42-L49

We could re-implement the code that we've been excluded from in Rust for static linking but that may be a lot of work.  If this were done we'd conditionally compile this code based on the `RUBY_STATIC` env.

Basically we need to get what isn't working in our loadpath during static builds to work.  There's no telling how simple or difficult this is with my current knowledge or Ruby's C code and C programming in general.

**Comment by felix-d** [comment]
Thanks for the writeup. My knowledge of Ruby internals and build process is quite limited as well, but this makes sense and definitely help me better understand the issue. I’ll keep investigating and keep you posted if I find anything.

**Comment by felix-d** [comment]
Something I've noticed: if I try to require another lib such as `zlib`, I get 

```
dyld: lazy symbol binding failed: Symbol not found: _rb_define_attr
  Referenced from: /opt/rubies/2.3.3/lib/ruby/2.3.0/x86_64-darwin16/zlib.bundle
  Expected in: flat namespace
```

However, `rb_define_attr` is not part of `internal.h` but `ruby.h` 🤔

**Comment by danielpclark** [comment]
> Something I've noticed: if I try to require another lib such as zlib

Did you make sure the flags you passed in were the absolute last item printed from `build.rs` ?  Because if the flags are early it messes things up.

**Comment by felix-d** [comment]
I may be missing something. I did not change `build.rs` to pass new flags. Also `RbConfig::CONFIG['LIBS']` returns `-lpthread -lgmp -ldl -lobjc` on my machine, if that's what you are referring to.

What I'm alluding to is [these failures](
https://travis-ci.org/danielpclark/rutie/builds/468453462?utm_source=github_status&utm_medium=notification) on the static builds [when trying to `require("zlib")`](https://github.com/danielpclark/rutie/pull/60). 

This is just an example to illustrate that it might not be related to `internal.h` as you mentioned, unless I'm mistaken somewhere, which is strongly possible.

**Comment by danielpclark** [comment]
No you're right. I was mistaken.  `LIBS` comes out to `"-lpthread -lgmp -ldl -lcrypt -lm "` on my machine.

**Comment by sandstrom** [comment]
Would it be helpful to look at how Helix solved this? (if they ever did)

https://github.com/tildeio/helix

My (limited) understanding is that Helix tried to solve basically the same problem that Rutie does. So maybe one could borrow techniques from them.

Or maybe reach out to @wagenet or @chancancode and see if they have any advice.
